File: xssfinder-java-executor/src/test/java/org/xssfinder/runner/SimpleMethodTraversalStrategyTest.java
package org.xssfinder.runner;

import org.junit.Before;
import org.junit.Test;
import org.xssfinder.Page;
import org.xssfinder.remote.TUntraversableException;
import org.xssfinder.remote.TraversalMode;

import java.lang.reflect.Method;

import static org.hamcrest.Matchers.instanceOf;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class SimpleMethodTraversalStrategyTest {
    private Method method;

    private SimpleMethodTraversalStrategy strategy;

    @Before
    public void setUp() throws Exception {
        method = RootPage.class.getMethod("goToSecondPage");
        strategy = new SimpleMethodTraversalStrategy();
    }

    @Test
    public void canSatisfyNormalTraversalMode() {
        // when
        boolean canSatisfy = strategy.canSatisfyMethod(method, TraversalMode.NORMAL);

        // then
        assertThat(canSatisfy, is(true));
    }

    @Test
    public void canSatisfySubmitTraversalMode() {
        // when
        boolean canSatisfy = strategy.canSatisfyMethod(method, TraversalMode.SUBMIT);

        // then
        assertThat(canSatisfy, is(true));
    }

    @Test
    public void traversingInvokesMethod() throws Exception {
        // given
        RootPage rootPage = new RootPage();

        // when
        TraversalResult traversalResult = strategy.traverse(rootPage, method);

        // then
        assertThat(traversalResult.getPage(), is(instanceOf(SecondPage.class)));
    }

    @Test(expected=RuntimeException.class)
    public void traversingMethodThatEncountersExceptionDoesNotCatchException() throws Exception {
        // given
        RootPage rootPage = new RootPage();
        method = RootPage.class.getMethod("raiseException");

        // when
        strategy.traverse(rootPage, method);
    }

    @Test(expected=TUntraversableException.class)
    public void traversingMethodWithArgumentsThrowsTUntraversableException() throws Exception {
        // given
        RootPage rootPage = new RootPage();
        method = RootPage.class.getMethod("withParameter", String.class);

        // when
        strategy.traverse(rootPage, method);
    }

    @SuppressWarnings("UnusedDeclaration")
    @Page
    private static class RootPage {
        public SecondPage goToSecondPage() {
            return new SecondPage();
        }
        public SecondPage raiseException() {
            throw new RuntimeException();
        }
        public SecondPage withParameter(String dummy) {
            return new SecondPage();
        }
    }

    @Page
    private static class SecondPage {}
}

File: xssfinder-php-executor/test/XssFinder/Runner/SimpleMethodTraversalStrategyTest.php
<?php

namespace XssFinder\Runner;

use PHPUnit_Framework_TestCase;
use ReflectionClass;
use ReflectionMethod;
use XssFinder\TraversalMode;

class SimpleMethodTraversalStrategyTest extends PHPUnit_Framework_TestCase
{
    /** @var SimpleMethodTraversalStrategy */
    private $_strategy;
    /** @var ReflectionMethod */
    private $_method;

    function setUp()
    {
        $rootPageClass = new ReflectionClass('SimpleMethodTraversalStrategy_TestPages\RootPage');
        $this->_method = $rootPageClass->getMethod('visitSomePage');

        $this->_strategy = new SimpleMethodTraversalStrategy();
    }

    function testCanSatisfyNormalTraversalMode() {
        // when
        $canSatisfy = $this->_strategy->canSatisfyMethod($this->_method, TraversalMode::NORMAL);

        // then
        assertThat($canSatisfy, is(true));
    }

    function testCanSatisfySubmitTraversalMode() {
        // when
        $canSatisfy = $this->_strategy->canSatisfyMethod($this->_method, TraversalMode::SUBMIT);

        // then
        assertThat($canSatisfy, is(true));
    }

    function testTraversingInvokesMethod()
    {
        // given
        $rootPage = new \SimpleMethodTraversalStrategy_TestPages\RootPage();

        // when
        $traversalResult = $this->_strategy->traverse($rootPage, $this->_method);

        // then
        assertThat($traversalResult->getPage(), is(anInstanceOf('SimpleMethodTraversalStrategy_TestPages\SecondPage')));
    }

    /**
     * @expectedException \Exception
     * @expectedExceptionMessage Test exception
     */
    function testTraversingMethodThatEncountersExceptionDoesNotCatchException()
    {
        // given
        $rootPage = new \SimpleMethodTraversalStrategy_TestPages\RootPage();
        $rootPageClass = new ReflectionClass('SimpleMethodTraversalStrategy_TestPages\RootPage');
        $this->_method = $rootPageClass->getMethod('raiseException');

        // when
        $this->_strategy->traverse($rootPage, $this->_method);
    }

    /**
     * @expectedException \XssFinder\TUntraversableException
     */
    function testTraversingMethodWithArgumentsThrowsTUntraversableException()
    {
        // given
        $rootPage = new \SimpleMethodTraversalStrategy_TestPages\RootPage();
        $rootPageClass = new ReflectionClass('SimpleMethodTraversalStrategy_TestPages\RootPage');
        $this->_method = $rootPageClass->getMethod('withParameter');

        // when
        $this->_strategy->traverse($rootPage, $this->_method);
    }
}

namespace SimpleMethodTraversalStrategy_TestPages;

class RootPage {
    public function visitSomePage() {
        return new SecondPage();
    }
    public function raiseException() {
        throw new \Exception('Test exception');
    }
    public function withParameter(\stdClass $param) {
        return new SecondPage();
    }
}

class SecondPage {}